File: data/adapters/openbench.ts
interface OpenBenchExtractionOptions {
  readonly condition: string;
  readonly excludedHarnesses?: readonly string[];
  readonly unreportedEffortHarnesses?: readonly string[];
}

function isRecord(value: unknown): value is Record<string, unknown> {
  return value !== null && typeof value === 'object' && !Array.isArray(value);
}

function text(value: unknown, label: string): string {
  if (typeof value !== 'string' || value.trim().length === 0) {
    throw new Error(`Invalid ${label}`);
  }

  return value;
}

function finiteNumber(value: unknown, label: string): number {
  if (typeof value !== 'number' || !Number.isFinite(value)) {
    throw new Error(`Invalid ${label}`);
  }

  return value;
}

function splitModelAndEffort(model: string): {
  readonly model: string;
  readonly effort: string | null;
} {
  const match = /^(.*)-(low|medium|high|xhigh)$/u.exec(model);

  return match
    ? { model: match[1] ?? model, effort: match[2] ?? null }
    : { model, effort: null };
}

function extractOpenBench(jsonl: string, options: OpenBenchExtractionOptions) {
  const excluded = new Set(['null', ...(options.excludedHarnesses ?? [])]);
  const unreportedEffort = new Set(options.unreportedEffortHarnesses ?? []);
  const lines = jsonl.split(/\r?\n/u).filter((line) => line.trim().length > 0);

  if (lines.length === 0) throw new Error('JSONL has no data rows');

  return lines.flatMap((line, index) => {
    let parsed: unknown;

    try {
      parsed = JSON.parse(line);
    } catch {
      throw new Error(`Invalid JSON on line ${index + 1}`);
    }

    if (!isRecord(parsed)) {
      throw new Error(`JSONL row ${index + 1} is not an object`);
    }

    const harness = text(parsed.harness, 'harness');
    if (excluded.has(harness)) return [];

    if (typeof parsed.success !== 'boolean') {
      throw new Error(`Invalid success on line ${index + 1}`);
    }

    const metrics: Record<string, number> = {
      solved: parsed.success ? 1 : 0,
      score:
        parsed.score === undefined
          ? parsed.success
            ? 1
            : 0
          : finiteNumber(parsed.score, 'score'),
      wall_time_s: finiteNumber(parsed.wall_time_s, 'wall time'),
    };

    if (parsed.tokens !== null && parsed.tokens !== undefined) {
      metrics.tokens = finiteNumber(parsed.tokens, 'tokens');
    }

    const configuration = splitModelAndEffort(text(parsed.model, 'model'));

    return [
      {
        model: configuration.model,
        effort: unreportedEffort.has(harness) ? null : configuration.effort,
        harness,
        benchmark: text(parsed.task, 'task'),
        condition: options.condition,
        trial: String(finiteNumber(parsed.trial, 'trial')),
        scoring: parsed.score === undefined ? 'success' : 'partial_score',
        metrics,
      },
    ];
  });
}

export { extractOpenBench };

File: research/src/eval_frontier/sources/adapters/openbench.py
"""Extract observations from OpenBench JSONL results."""

from __future__ import annotations

import json
import re
from typing import Any


def _text(value: Any, label: str) -> str:
    if not isinstance(value, str) or not value.strip():
        raise ValueError(f"Invalid {label}")
    return value


def extract(
    content: str,
    condition: str,
    excluded: set[str] | None = None,
    no_effort: set[str] | None = None,
) -> list[dict[str, Any]]:
    excluded = {"null", *(excluded or set())}
    no_effort = no_effort or set()
    rows: list[dict[str, Any]] = []
    for line_no, line in enumerate(content.splitlines(), 1):
        if not line.strip():
            continue
        try:
            item = json.loads(line)
        except json.JSONDecodeError as exc:
            raise ValueError(f"Invalid JSON on line {line_no}") from exc
        harness = _text(item.get("harness"), "harness")
        if harness in excluded:
            continue
        if not isinstance(item.get("success"), bool):
            raise ValueError(f"Invalid success on line {line_no}")
        score = item.get("score", 1 if item["success"] else 0)
        if not isinstance(score, (int, float)) or isinstance(score, bool):
            raise ValueError("Invalid score")
        wall = item.get("wall_time_s")
        if not isinstance(wall, (int, float)) or isinstance(wall, bool):
            raise ValueError("Invalid wall time")
        model_raw = _text(item.get("model"), "model")
        match = re.match(r"^(.*)-(low|medium|high|xhigh)$", model_raw)
        model, effort = (match.group(1), match.group(2)) if match else (model_raw, None)
        trial_raw = item.get("trial")
        trial = str(trial_raw) if trial_raw is not None else None
        metrics: dict[str, float] = {
            "solved": int(item["success"]),
            "score": score,
            "wall_time_s": wall,
        }
        if item.get("tokens") is not None:
            if not isinstance(item["tokens"], (int, float)) or isinstance(item["tokens"], bool):
                raise ValueError("Invalid tokens")
            metrics["tokens"] = item["tokens"]
        rows.append(
            {
                "_source_line": line_no,
                "model": model,
                "effort": None if harness in no_effort else effort,
                "harness": harness,
                "benchmark": _text(item.get("task"), "task"),
                "condition": condition,
                "trial": trial,
                "scoring": "success" if "score" not in item else "partial_score",
                "metrics": metrics,
            }
        )
    if not rows:
        raise ValueError("JSONL has no data rows")
    return rows
